### ISSUE TITLE
CodeSandbox+TypeScript: Optional chaining operator in JSX expressions

### ISSUE THREAD

**Issue by rturnq** [ryansolid/solid#150]
It looks like using the optional chaining operator (?.) in a JSX expression doesn't work as I would expect - it will render initially but never update.

For example given a state

```typescript
const [state, setState] = createState<{
  lastUpdated?: Date
}>({})
```

and the following JSX
```tsx
<div>{state.lastUpdated?.toISOString()}</div>
```

it will compile to something like
```javascript
(0, _dom.insert)(_el$10, function () {
  var _c$ = (0, _dom.wrapCondition)(function () {
    return (_a = state.lastUpdate) === null || _a === void 0;
  });

  return function () {
    return _c$() ? void 0 : _a.toISOString();
  };
}());
```
Which only runs once, though I don't really understand why.

Alternatively, if `lastUpdated` is not optional
```tsx
<div>{state.lastUpdated.toISOString()}</div>
```

or I use TypeScript's non-null assertion (!.)
```tsx
<div>{state.lastUpdated!.toISOString()}</div>
```

it will compile to something like
```javascript
(0, _dom.insert)(_el$7, function () {
  return state.lastUpdate.toISOString();
});
```

Which works as expected.

Finally, I discovered by accident that one can get around the issue by making the expression a function. For instance
```tsx
<div>{() => state.lastUpdated?.toISOString()}</div>
```

will compile to something like
```javascript
(0, _dom.insert)(_el$4, function () {
  var _a;

  return (_a = state.lastUpdate) === null || _a === void 0 ? void 0 : _a.toISOString();
})
```

Is usage of this form (an expression to be wrapped in a function) intentional and supported or does it just happen to work? Can compilation be changed to allow the non-function version of the expression to yield the same result?

**Comment by ryansolid** [comment]
Interesting. I actually used optional chaining to great success in my Suspense examples. Like this one: https://codesandbox.io/s/solid-suspenselist-eorvk. Basically if it wasn't working the data would never load in. The conditionalWrap should also work correctly eyeballing it.. It creates a Memo that updates on `state.lastUpdate` since that always evals in the course of evaluating that computation.

The wrapped in a function is intentional because I need to handle dynamic inserts. However it should be unnecessary in this case (it basically creates an additional computation). I suppose I can try to make a minimal reproduction case.. Or if you have one? As to my knowledge this feature already works.

**Comment by rturnq** [comment]
Here is a minimal example https://codesandbox.io/s/solidissue-150-27ex6.

After thinking about the compiled code a little more it makes sense that it doesn't work in this case. `_c$` is a memo that returns the values true or false depending on if `lastUpdated` is null or undefined. I am never setting `lastUpdated` to null or undefined, it is only undefined when the state is initialized. Each time `lastUpdated` is set to a new date object, the memoized code is re-executed (which can be verified by a breakpoint or console log) but because it returns true again, no change has occurred and dependents are (correctly) not updated.

**Comment by ryansolid** [comment]
Oh you are right. For some reason I was thinking it was using the "||" like a ternary but it's just two booleans. Hmm yeah this is interesting. The problem is that the property being accessed after the chaining variable isn't reactive since it's a date object. If this was nested reactive data there would be no problem. Since in my case:

```jsx
// props.user?.name becomes:
(0, _dom.insert)(_el$10, function () {
  var _c$ = (0, _dom.wrapCondition)(function () {
    return (_a = state.user) === null || _a === void 0;
  });

  return function () {
    return _c$() ? void 0 : _a.name;
  };
}());
```
So `_a.name` is still reactive. So correctly when moving from false to true it retriggers which lets it track `_a.name` which from that point looks for name changes. However in this case the `_a` is not a reactive state object it's a date. So it never tracks. Once the date is set the condition ensures it so. Why I do that is I can't always guarantee the left side of the comparison won't change multiple times. The `state.count > 0` problem. Even in the case of changing users here you might not want to recreate the whole thing and just say pass a different name down in a prop further down.

Like consider the difference between:
```jsx
<Show when={state.user}>...</Show>
<Show when={!!state.user}>...</Show>
```
It's unfortunate that the transpilation takes this away from you. Since there is a difference here between:
```js
state.date == undefined ? void 0 : state.date.toISOString()
// and
(a = state.date) == undefined ? void 0 : a.toISOString()
```
The former would be fine but the latter is not if `a.toISOString()` is not reactive. Which it is not. The reason the function works is I only do condition analysis directly in template expressions. Hoisting (or inlining) a function bypasses it.

So this is a tricky one. I did consider only wrapping conditionals with comparison operators that were not pure equivalents but as I said that takes the power from the end user to make the decision (the show example). My wrappers don't assume booleans so you can also just return values that can change.  I think the grammar and heuristics makes sense in general this only is an issue if all 3 of the following are true:
1. No native optional chaining support so it is transpiled.
2. The object with the optional chaining is not reactive.
3. The property being accessed/called is not reactive. 

In short.. I think Dates are realistically the only place this would generally happen in a view. It isn't a permanent problem nor one that you would hit that often, but it is not something that could be easily determined statically. So I don't think I can fix this either. It's one of those special notes around Date Objects I think.

**Comment by rturnq** [comment]
I mostly follow what you are saying.

I'm not sure I understand what you are getting at with the `<Show />` example. `when` is expecting a boolean expression an is ultimately memoized as `!!props.when` within the Show component. Passing in an `object | undefined` happens to work fine in JavaScript and gives you a type mismatch error in TypeScript. Passing in an object coerced to a boolean should have the same end result, right?

As for the issue at hand, I disagree that Date objects are special to this issue and even that they are the most likely way someone would encounter it. Here are some other examples that fail to track: https://codesandbox.io/s/solidissue-150-t03om

Some of the examples are pretty contrived and I would likely not use several of them in the real world; however, it appears that any instance method called on an object or primitive will cause this. So something as innocent as
```jsx
<h3>{user?.name.toLowerCase()}</h3>
```
may land you in this situation.

I understand that this might not be fixable from your end through static analysis without taking away from the end user and am fine with that. What is your recommendation for a solution on my end? Should I wrap these expressions in a function or always memoize them myself?

**Comment by ryansolid** [comment]
I see. I was focusing on the function calls on the end, but really matters at which point it needs to listen at. The reason your examples where it's in the first position like:
```jsx
<span>{state.account?.total.toFixed(2)}</span>
```
aren't working since you are replacing account with `setState`.  So I see what you are saying. If the reactive change happens before the optional chaining operator it gets lost. This is so unfortunate since it's not the syntax but the way it transpiles. When this is natively supported there will be no issue. I almost feel like I need to write my own babel plugin to fix this and maybe package it in with Babel Preset Solid as an option. If it didn't do that assignment this would work fine. I actually think this might be the only option.

You can use the function for sure for inserts but that won't work for props, You can hoist it or do a self calling function in both cases but it's nasty.  Yeah I think I need to write a Babel plugin.

**Comment by ryansolid** [comment]
Oh you are right about `Show` I must have changed that at some point(apparently 5 months ago). Originally it didn't force the boolean so the difference there was one would trigger every time the user changed and the other only when there was any user. Right now I'm surprised I went this way but it was probably to simplify things. I think I did the same on conditionals until I realized an awkwardness on `||` statements where you want to return the condition.

**Comment by ryansolid** [comment]
@rturnq Yeah putting it after works. Which actually means if this was anything but parcel and code sandbox it would work.

I don't see the optional chaining plugin so I assume it's part of `preset-env` now. Which is good other than the way babel works it resolves plugins in this order:

1. First Plugin to Last Plugin
2. Last Preset to first Preset

Any sane configuration always puts `preset-env` first atleast my experience, but either parcel or codesandbox secretly push it on the end for some reason which means there is no way to run a preset before it. Even if you manually add it still gets executed after all plugins but before other presets. Which means that I don't really have any choice there, but if you make your own rollup or webpack config, this whole thing isn't even an issue.

**Comment by rturnq** [comment]
Thanks for digging into this. Yep, it looks like it is coming from TypeScript when targeting anything less than ESNext: https://www.typescriptlang.org/play/?target=3#code/JYOwLgpgTgZghgYwgAgMpjpZBvAUMg5AGzgGcwBVABwBNMIaB+ALmQBF7cBfXXBAexDlk5eq3T1kAXhw9coyADoS5anUhNFYfgElUAeXRRQAcwAUASiA

Setting the target to ES2020 or ESNext preserves the syntax so it can be handled further down the compilation pipeline. I'm not sure what side-effects this may have though.

I'm glad you don't have to write a babel plugin just for this.

**Comment by ryansolid** [comment]
Yeah I was using babel-preset-typescript on my stuff so it was handling it later as well. I imagine in babel-preset-env which as I said is always last to resolve. Ok.. well I guess realistically there is nothing to fix here. We just have to be wary of typescripts doing transforms ahead. Which generally is a good suggestion when you have babel in the toolchain anyway. I'm glad this wasn't more serious. I really like having optional chaining. It really highlights Solid's ability to push reactivity to the leaves. That Suspense example I linked is gold. The child Component doesn't even have to be aware of Suspense. As long as it knows that the data isn't guaranteed to be there. A quick optional chaining and that's all you need. Even though that expression triggers Suspense it comes down to a simple is it there or not.

**Comment by ryansolid** [comment]
Nothing to really do here other than make a bit of a note. Since you have to use Solid with Babel anyway it is sort of setting a course to suggest we can't use TS on its own anyway. Hmm.. I don't have a suitable place to put this note yet though. It is super specific, this is more of how to on getting started with TypeScript. But even then this is too detailed. Its almost better to provide a sample config. I'm gathering I'm going to close this soon unless there is something else to be done here.

**Comment by rturnq** [comment]
I agree. The only thing I could think is a quick note for getting started with TypeScript that just suggests targeting ESNext if you are using `tsc` to prevent it from doing transformations before it gets to Babel. Feel free to close and remove the bug label.

**Comment by ryansolid** [comment]
Yeah I will do that.. in process of writing new docs. So hope to address it there.
